// @sequelize/postgres/lib/_internal/connection-options.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var connection_options_exports = {};
__export(connection_options_exports, {
  BOOLEAN_CONNECTION_OPTION_NAMES: () => BOOLEAN_CONNECTION_OPTION_NAMES,
  CONNECTION_OPTION_NAMES: () => CONNECTION_OPTION_NAMES,
  NUMBER_CONNECTION_OPTION_NAMES: () => NUMBER_CONNECTION_OPTION_NAMES,
  STRING_CONNECTION_OPTION_NAMES: () => STRING_CONNECTION_OPTION_NAMES
});
module.exports = __toCommonJS(connection_options_exports);
var import_utils = require("@sequelize/utils");
const STRING_CONNECTION_OPTION_MAP = {
  application_name: void 0,
  client_encoding: void 0,
  database: void 0,
  host: void 0,
  options: void 0,
  password: void 0,
  user: void 0
};
const STRING_CONNECTION_OPTION_NAMES = (0, import_utils.getSynchronizedTypeKeys)(
  STRING_CONNECTION_OPTION_MAP
);
const BOOLEAN_CONNECTION_OPTION_MAP = {
  binary: void 0,
  keepAlive: void 0,
  ssl: void 0,
  statement_timeout: void 0
};
const BOOLEAN_CONNECTION_OPTION_NAMES = (0, import_utils.getSynchronizedTypeKeys)(
  BOOLEAN_CONNECTION_OPTION_MAP
);
const NUMBER_CONNECTION_OPTION_MAP = {
  port: void 0,
  statement_timeout: void 0,
  query_timeout: void 0,
  keepAliveInitialDelayMillis: void 0,
  idle_in_transaction_session_timeout: void 0,
  connectionTimeoutMillis: void 0,
  lock_timeout: void 0
};
const NUMBER_CONNECTION_OPTION_NAMES = (0, import_utils.getSynchronizedTypeKeys)(
  NUMBER_CONNECTION_OPTION_MAP
);
const CONNECTION_OPTION_NAMES = (0, import_utils.getSynchronizedTypeKeys)({
  ...STRING_CONNECTION_OPTION_MAP,
  ...BOOLEAN_CONNECTION_OPTION_MAP,
  ...NUMBER_CONNECTION_OPTION_MAP,
  stream: void 0
});
